Implement JSONRPC 2.0 Server

hgoldstein95 opened the pull request:
Implement server for Tourist that will allow for significant improvements to the VSCode front-end, and that will also enable easier implementation of other front-ends. As of now, there are a few pieces remaining before this should be merged:

- [x] Close issues on the relevant milestone.
- [ ] @J-Fields should review the change, probably with guidance from @hgoldstein95 on more complex corners of the server.

hgoldstein95 committed:
Add JSON RPC 2.0 server, basic API stable

hgoldstein95 committed:
Split interface into separate file

hgoldstein95 committed:
Documentation improvements

hgoldstein95 committed:
Broke IO out into separate service for async

hgoldstein95 committed:
Add a way to link stops

hgoldstein95 committed:
More implementation for server

hgoldstein95 committed:
Implement deleting a tour

hgoldstein95 committed:
Finish base implementation

Add repl for ease of testing

hgoldstein95 committed:
Add tests of Tourist object

hgoldstein95 committed:
More tests

hgoldstein95 committed:
Refactor index, tests flaky

hgoldstein95 committed:
Fix flaky tests, factor index into trait

hgoldstein95 committed:
Add unlink stop

hgoldstein95 committed:
Simplify error model for easier integration with failure model

hgoldstein95 committed:
Unify error handling between commands

hgoldstein95 committed:
Added error handling to FileIndex

hgoldstein95 committed:
Factor index into module

hgoldstein95 committed:
Start the process of adding more config options

hgoldstein95 committed:
Started working on default tours

hgoldstein95 committed:
Finish default tour init

hgoldstein95 committed:
Added better handling of cfg failure

Config path can now be configured with dotenv

hgoldstein95 committed:
Add lots of new tests, including some PBT

hgoldstein95 committed:
Add more testing to core features, fixed bugs

hgoldstein95 committed:
Improved testing, added framework for integration tests

See `server.rs` for integration test framework. We need to flesh this out ASAP

hgoldstein95 committed:
Not sure why this was checked in to begin with

J-Fields commented on src/command/serve/interface.rs:
```diff
@@ -0,0 +1,206 @@
+    ) -> JsonResult<()>;
+
+    /// Unlink a tour stop from another tour or tour stop.
+    #[rpc(name = "link_stop")]
```
Typo I think?

hgoldstein95 replied:
Yep, just pushed a fix.

J-Fields commented:
Just so you're aware, few test failures on windows, haven't looked into it yet
```
test command::serve::tests::create_stop_test ... FAILED
test command::serve::tests::locate_stop_test ... FAILED
test command::serve::tests::refresh_tour_test ... FAILED
test types::path::relative::tests::components_consistent_with_path ... FAILED
test command::serve::tests::index_repository_test ... FAILED
```

hgoldstein95 committed:
Fix typo for "unlink_stop" method

hgoldstein95 commented:
I'd have to guess that those have something to do with why you can't load any stops lol. Do you have the output of the failures themselves?

J-Fields commented:
```
---- command::serve::tests::create_stop_test stdout ----
thread 'command::serve::tests::create_stop_test' panicked at 'called `Option::unwrap()` on a `None` value', src\libcore\option.rs:347:21
stack backtrace:
   0: backtrace::backtrace::trace_unsynchronized
             at C:\Users\VssAdministrator\.cargo\registry\src\github.com-1ecc6299db9ec823\backtrace-0.3.29\src\backtrace\mod.rs:66
   1: std::sys_common::backtrace::_print
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b\/src\libstd\sys_common\backtrace.rs:47
   2: std::sys_common::backtrace::print
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b\/src\libstd\sys_common\backtrace.rs:36
   3: std::panicking::default_hook::{{closure}}
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b\/src\libstd\panicking.rs:200
   4: std::panicking::default_hook
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b\/src\libstd\panicking.rs:211
   5: std::panicking::rust_panic_with_hook
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b\/src\libstd\panicking.rs:477
   6: std::panicking::continue_panic_fmt
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b\/src\libstd\panicking.rs:384
   7: std::panicking::rust_begin_panic
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b\/src\libstd\panicking.rs:311
   8: core::panicking::panic_fmt
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b\/src\libcore\panicking.rs:85
   9: core::panicking::panic
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b\/src\libcore\panicking.rs:49
  10: core::option::Option<tourist::types::path::absolute::AbsolutePathBuf>::unwrap<tourist::types::path::absolute::AbsolutePathBuf>
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b\src\libcore\macros.rs:12
  11: tourist::command::serve::tests::create_stop_test
             at .\src\command\serve\tests.rs:359
  12: tourist::command::serve::tests::create_stop_test::{{closure}}
             at .\src\command\serve\tests.rs:354
  13: core::ops::function::FnOnce::call_once<closure,()>
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b\src\libcore\ops\function.rs:231
  14: <unknown>
```

hgoldstein95 committed:
Add slog logger to help track down errors

Logging happens when "-v" is passed to any subcommand. The logs are sent to
.tourist.log, and overwrite on restart.

hgoldstein95 commented:
Oh, I'm dumb and those tests aren't portable at all. Let me fix that.

J-Fields commented on repl/tourist.py:
```diff
@@ -0,0 +1,90 @@
+    def forget_tour(self, tour_id):
+        return self.__rpc("forget_tour", [tour_id])
+
+    def create_stop(self, tour_id, title, path, line):
```
Can you add an `index` param? Need it to enable undoing deletions.

hgoldstein95 replied:
Index is supposed to be internal state of the server. Why do you need that in order to undo deletions?

J-Fields replied:
If I delete the 2nd stop in a 3-stop tour, how do I undo that operation?

hgoldstein95 replied:
Should probably just provide a basic undo log for delete operations, but now I also get what you meant. I read index as "repo index" and was confused about what that had to do with anything. I totally forgot to add move and relocate commands for stops, so that's important. Those will have index params lol

J-Fields commented:
Any reason there's no `move_stop`?

hgoldstein95 commented:
Oversight :/ sorry this is such a mess. Just keep the tickets/tasks coming and I'll get around to it.

J-Fields commented:
No worries, this is a much nicer interface than the typescript library's

hgoldstein95 committed:
Add move_stop and reorder_stop to interface

hgoldstein95 committed:
Fixed tests to work generically

hgoldstein95 commented:
@J-Fields I think I fixed the tests that fail on Windows, but let me know if that's not the case.

hgoldstein95 committed:
Allow stops to be broken

- Add `broken` field to stops
- Set/unset where appropriate

hgoldstein95 committed:
Actually use editable flag properly

hgoldstein95 committed:
Call refresh when opening a tour

hgoldstein95 committed:
Fix bug with resolve

hgoldstein95 committed:
Default tours now have paths set appropriately

hgoldstein95 committed:
Added some logging and cleaned some things up

hgoldstein95 committed:
Cleaned up error handling, added error codes

hgoldstein95 committed:
Add reload_tour to interface

When calling `set_tour_edit` with `false`, `reload_tour` is called automatically.

hgoldstein95 committed:
Split `set_tour_edit` into `freeze_tour` and `unfreeze_tour`

hgoldstein95 committed:
Add safeguards to prevent some operations when repo is dirty

The user should not be able to create or move a stop if the tour has
underlying repositories that have local changes.

hgoldstein95 committed:
When removing stop, also remove any repository that is unused

hgoldstein95 committed:
Add tourist_ref(_mut) macros to clean up tour and stop lookup

hgoldstein95 committed:
Huge refactor to pull logic out of json api

hgoldstein95 committed:
Additions to VCS to enable checking out tours

Bump `git2` version to 0.10

hgoldstein95 committed:
Finish implementing checkout_for_tour

hgoldstein95 committed:
Pulled thread safety concerns out of engine, pushed into jsonrpc

Previously, the `Engine` needed to be thread-safe because it implemented
`TouristRpc`. Now, we just implement `TouristRpc` for `Arc<RwLock<Engine>>` and
don't worry about thread safety in the Engine

hgoldstein95 committed:
Removed `save_all`

The file manager API is a little bit clunky already. I'd rather not have to
maintain save_all, especaially because I don't see a good reason to use it.

hgoldstein95 committed:
Finished untangling the sync story in serve

There were still some artifacts of the Send + Sync engine that held over in the
tests, the Index trait, and the VCS trait. I just cleaned those up.

hgoldstein95 committed:
Add utility macro for creating hashmaps, start adding logging

hgoldstein95 committed:
Removed `generator` from Tour

It's not necessary anymore, and it confused me a couple of times. This might
make interop with old versions of the VSCode extension weird, but I don't
actually think it's going to be an issue.

hgoldstein95 committed:
More logging changes, refactored refresh

Refresh used to run a diff for each stop -- now it only runs one per repository.
In addition, refresh used to be able to take a commit version as a parameter;
this was totally broken for any tour with more than one repository.

hgoldstein95 committed:
Added logging all over the place

hgoldstein95 committed:
Add `up_to_date` flag to tour view

hgoldstein95 committed:
Replace unwrap with expect wherever applicable

There are only a few instances of unwrap now -- all are in engine/tests.rs. The
other unwraps were either replaced with more appropriate error handling or at
least replaced with expect with some sort of identifying message.

hgoldstein95 committed:
Change `checkout_for_tour` to roll back if it fails

Now, if `checkout_for_tour` fails to check out some repo, the others will be
reset to their original commits.

hgoldstein95 committed:
command/serve: Use parking_lot version of RwLock.

The `parking_lot` crate has better locking primitives than the standard library.
Chaning to `parking_lot::RwLock` obviates the need to check for a poisoned lock.

hgoldstein95 committed:
project: Fix clippy lints introduced with new version

hgoldstein95 committed:
engine/tests: Fix issue #21, tests failing on WSL

Tests were failing on WSL because there was no downloads directory. We now use
`std::env::current_dir` instead of the downloads dir.

hgoldstein95 committed:
engine: Fix #22. Tours are now created in edit mode.
